odoo/models.py

```diff
@@ -3760,74 +3760,77 @@ def unlink(self):
-            if ir_model_data_unlink:
-                ir_model_data_unlink.unlink()
-            if ir_attachment_unlink:
-                ir_attachment_unlink.unlink()
-            # DLE P93: flush after the unlink, for recompute fields depending on
-            # the modified of the unlink
-            self.env.flush_all()
+        cr = self._cr
+        Data = self.env['ir.model.data'].sudo().with_context({})
+        Defaults = self.env['ir.default'].sudo()
+        Property = self.env['ir.property'].sudo()
+        Attachment = self.env['ir.attachment'].sudo()
+        ir_model_data_unlink = Data
+        ir_attachment_unlink = Attachment
+
+        for sub_ids in cr.split_for_in_conditions(self.ids):
+            records = self.browse(sub_ids)
+
+            # Check if the records are used as default properties.
+            refs = ['%s,%s' % (self._name, i) for i in sub_ids]
+            if Property.search([('res_id', '=', False), ('value_reference', 'in', refs)], limit=1):
+                raise UserError(_('Unable to delete this document because it is used as a default property'))
+
+            # Delete the records' properties.
+            Property.search([('res_id', 'in', refs)]).unlink()
+
+            # mark fields that depend on 'self' to recompute them after 'self' has
+            # been deleted (like updating a sum of lines after deleting one line)
+            with self.env.protecting(self._fields.values(), records):
+                self.modified(self._fields, before=True)
+
+            query = "DELETE FROM %s WHERE id IN %%s" % self._table
+            cr.execute(query, (sub_ids,))
+
+            # Removing the ir_model_data reference if the record being deleted
+            # is a record created by xml/csv file, as these are not connected
+            # with real database foreign keys, and would be dangling references.
+            #
+            # Note: the following steps are performed as superuser to avoid
+            # access rights restrictions, and with no context to avoid possible
+            # side-effects during admin calls.
+            data = Data.search([('model', '=', self._name), ('res_id', 'in', sub_ids)])
+            if data:
+                ir_model_data_unlink |= data
+
+            # For the same reason, remove the defaults having some of the
+            # records as value
+            Defaults.discard_records(records)
+
+            # For the same reason, remove the relevant records in ir_attachment
+            # (the search is performed with sql as the search method of
+            # ir_attachment is overridden to hide attachments of deleted
+            # records)
+            query = 'SELECT id FROM ir_attachment WHERE res_model=%s AND res_id IN %s'
+            cr.execute(query, (self._name, sub_ids))
+            attachments = Attachment.browse([row[0] for row in cr.fetchall()])
+            if attachments:
+                ir_attachment_unlink |= attachments.sudo()
```

Comment:
useless sudo ? it take the env of `ir_attachment_unlink`, no? 
(not link to your change, but if you rewrite the method, you can do this change)